File: fsharp/bank-account/BankAccount.fs
module BankAccount
open System

type Account = private { 
    mutable balance: float; 
    isOpen: bool; 
    lock: Object;
}

let mkBankAccount() = { 
    balance = 0.0; 
    isOpen = false; 
    lock = new Object();
}

let openAccount account = { account with isOpen = true }

let closeAccount account = { account with isOpen = false }

let getBalance account = if not account.isOpen then None else Some account.balance

let updateBalance change account =
    lock account.lock (fun () -> account.balance <- account.balance + change)
    account

File: csharp/bank-account/BankAccount.cs
using System;

public class BankAccount
{
    private bool isOpen = false;
    private float balance;

    private object lockObject = new object();

    public void Open() => isOpen = true;

    public void Close() => isOpen = false;

    public float Balance 
    { 
        get 
        { 
            if(!isOpen)
                throw new InvalidOperationException("This account is closed");
            return balance;
        }
        private set { balance = value; }
    }

    public void UpdateBalance(float change)
    {
        lock (lockObject)
        {
            Balance += change;
        }
    }
}
